src/Codecs.chpl

```diff
@@ -0,0 +1,110 @@
+module Codecs {
+  use CTypes;
+  use idna;
+  use iconv;
+
+  use AryUtil;
+  
+  proc encodeStr(obj: c_ptr(uint(8)), inBufSize: int, outBufSize: int, toEncoding: string = "UTF-8", fromEncoding: string = "UTF-8"): [] uint(8) throws {
+    if toEncoding == "IDNA" {
+      if outBufSize == 1 {
+        var ret: [0..0] uint(8) = 0x00;
+        return ret;
+      }
+      var cRes: c_string;
+      var rc = idn2_to_ascii_lz(obj:c_string, cRes, 0);
+      if (rc != IDNA_SUCCESS) {
+        throw new Error("Encode failed");
+      }
+      var tmp = cRes: c_ptr(uint(8));
+      var ret: [0..#outBufSize] uint(8);
+      for i in ret.domain do ret[i] = (tmp+i).deref();
```

Comment:
is there a reason this is a `for` loop instead of a `forall`? each loop iteration seems independent of the others to me, but I've never used `deref()` before

Reply:
No, I think that can probably be a forall; good catch.